src/hotspot/share/classfile/classLoaderDataGraph.cpp

```diff
@@ -520,7 +527,8 @@ bool ClassLoaderDataGraph::do_unloading() {
       prev->set_next(data);
     } else {
       assert(dead == _head, "sanity check");
-      _head = data;
+      // The GC might be walking this concurrently
+      Atomic::store(&_head, data);
```

Comment:
If you are using `load_acquire` then surely this must be a `release-store` as they need to be paired.

Reply:
The load_acquire synchronizes with release_store on insertions. This relaxed unlinking store, doesn't publish anything new that the load_acquire is interested in. It rolls back the head to a CLD that was installed at some earlier point with release_store, meaning it has already been published safely before.

Reply:
Thanks for clarifying